proof-of-spacetime.md

```diff
@@ -0,0 +1,200 @@
+└───────┘                                     └──────────────────┘└───────┘
+                                                        ▲
+
+                                                        │
+                                                                       ┌ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+                                                        └ ─ ─ ─ ─ ─ ─ ─ Randomness Input
+                                                                       └ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+```
+
+TODO: Add post submission to the diagram.
+
+
+
+
+### High Level API
+
+#### Generation
+
+```go
+func GeneratePoSt(sectorSize BytesAmount, sectors []commR) (PoStProof, FaultSet) {
+    // Generate the Merkle Inclusion Proofs + Faults
+
+    inclusionProofs := []
+    faults := NewFaultSet()
+
+    for n in 0..POST_CHALLENGES_COUNT {
+        attempt := 0
+        'inner: for {
+            challenge := DerivePoStChallenge(seed, n, faults, attempt)
+            sector := challenge % sectorSize
+            // check if we landed in a previously marked faulty sector
+            if !faults.Contains(sector) {
+                attempt += 1
+                continue
+            }
+
+            challenge_value = challenge / sectorSize
+            inclusionProof, isFault := GenerateMerkleInclusionProof(sector, challenge_value)
+            if isFault {
+                // faulty sector, generate a new challenge
+                faults.Add(sector)
+                attempt += 1
+            } else {
+                // no fault, move on to the next challenge
+                inclusionProofs[n] = inclusionProof
+                break 'inner
+            }
+        }
+    }
+
+    // Generate the snark
+    challenges := DerivePoStChallenges(sectorSize, seed, faults)
+
+    snark_proof := GeneratePoStSnark(sectorSize, challenges, sectors, inclusionProofs)
+
+    proof := PoStProof {
+        snark: snark_proof
+    }
+
+    return proof, faults
+}
+```
+
+#### Verification
+
+
+```go
+func VerifyPoSt(sectorSize BytesAmount, sectors []commR, seed []byte, proof PoStProof, faults FaultSet) bool {
+    challenges := DerivePoStChallenges(sectorSize, seed, faults)
+
+    // Verify snark
+    VerifyPoStSnark(sectorSize, challenges, sectors)
+}
+```
+
+
+#### Types
+
+```go
+type PoStProof struct {
+    snark []byte
+}
+```
+
+#### Challenge Derivation
+
+```go
+// Derive the full set of challenges for PoSt.
+func DerivePoStChallenges(sectorSize BytesAmount, seed []byte, faults FaultSet) [POST_CHALLENGES_COUNT][]byte {
+    challenges := []
+
+    for n in 0..POST_CHALLENGES_COUNT {
+        attempt := 0
+        'inner: for {
+            challenge := DerivePoStChallenge(seed, n, faults, attempt)
+
+            // check if we landed in a faulty sector
+            sector := challenge % sectorSize
+            if !faults.Contains(sector) {
+                // Valid challenge
+                challenges[n] = challenge
+                break 'inner
```

Comment:
is there a simpler way that define how this works without break?

it's very difficult to reason through this

Reply:
fixed